[GTK] Implement the X25519 algorithm

javifernandez opened the pull request:
#### 0dab8bb6c4d71fb6c3072859b974642a6cf7487a
<pre>
[GTK] Implement the X25519 algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=249017">https://bugs.webkit.org/show_bug.cgi?id=249017</a>

Reviewed by Youenn Fablet and Žan Doberšek.

This change implements all the X25519&apos;s operations defined in the
WebCrypto Secure Curves specification. These operations include the
key generation, import/export in spki, pkcs8 and jwk formats and the
deriveKey and deriveBits methods that are the key operations of this
algorithm.

The main WebCrypto API implementation of the GTK+ port is based on
the Libgcrypt primitives. There is an ongoing effort to implement it
in OpenSSL (see bug 248980 for details) but it won&apos;t be an option for
the GTK+ port in the short term.

A new utility class has been added to the GCrypt based implementation
of the WebCrypto API, named as GCryptRFC7748. It implements the a function
to perform scalar multiplication on the Montgomery form of the Curve25519
and Curve488 curves. However, it only defines the X25519 function since
we don&apos;t support the X488 algorithm for now.

The feature is implemented behind a runtime flag, hence this PR adds a
new item in the UnifiedWebPreferences file as &apos;WebCryptoX25519Enabled&apos;
with &apos;testable&apos; status.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/cfrg_curves_bits.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/cfrg_curves_bits.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/cfrg_curves_keys.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/cfrg_curves_keys.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/failures_X25519.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/failures_X25519.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/successes_X25519.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/successes_X25519.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_X25519.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_X25519.https.any.worker-expected.txt: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::write):
(WebCore::CloneDeserializer::read):
* Source/WebCore/crypto/CryptoAlgorithmIdentifier.h:
* Source/WebCore/crypto/CryptoAlgorithmParameters.h:
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::isX25519Enabled):
(WebCore::normalizeCryptoAlgorithmParameters):
(WebCore::isSupportedExportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp: Added.
(WebCore::CryptoAlgorithmX25519::create):
(WebCore::CryptoAlgorithmX25519::identifier const):
(WebCore::CryptoAlgorithmX25519::generateKey):
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):
(WebCore::CryptoAlgorithmX25519::deriveBits):
(WebCore::CryptoAlgorithmX25519::importKey):
(WebCore::CryptoAlgorithmX25519::exportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.h: Added.
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmRegistryGCrypt.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmX25519GCrypt.cpp: Added.
(WebCore::gcryptDerive):
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):
* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::CryptoKeyOKP::isPlatformSupportedCurve):
(WebCore::CryptoKeyOKPImpl::supportedAlgorithmIdentifier):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::exportSpki const):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::exportPkcs8 const):
(WebCore::CryptoKeyOKP::generateJwkX const):
* Source/WebCore/crypto/gcrypt/GCryptRFC7748.cpp: Added.
(WebCore::xImpl):
(WebCore::X25519):
* Source/WebCore/crypto/gcrypt/GCryptRFC7748.h: Added.
* Source/WebCore/crypto/gcrypt/GCryptUtilities.h:
* Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
(WebCore::CryptoKeyOKP::importRaw):
(WebCore::CryptoKeyOKP::importJwk):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::platformExportRaw const):
* Source/WebCore/crypto/keys/CryptoKeyOKP.h:
* Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
* Source/WebCore/crypto/parameters/CryptoAlgorithmX25519Params.h: Added.
* Source/WebCore/crypto/parameters/X25519Params.idl: Added.
* Source/WebCore/platform/SourcesGCrypt.txt:

Canonical link: <a href="https://commits.webkit.org/267602@main">https://commits.webkit.org/267602@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/412a36bb7bb8c99e2f2893cc56459410205d57fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16958 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17743 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18742 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15876 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17149 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20514 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17421 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18109 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17158 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17506 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14686 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19556 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14749 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15381 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22107 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14635 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15733 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15548 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19869 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/16184 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16155 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13687 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18517 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15317 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4319 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4089 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19682 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19741 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15986 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4172 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

zdobersek commented on Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
```diff
@@ -116,20 +116,23 @@ RefPtr<CryptoKeyOKP> CryptoKeyOKP::importJwk(CryptoAlgorithmIdentifier identifie
+
+    if (!keyData.d.isNull()) {
+        auto d = base64URLDecode(keyData.d);
+        if (!d || !platformCheckPairedKeys(identifier, namedCurve, WTFMove(*d), WTFMove(*x)))
```
The vectors shouldn't be moved into the `platformCheckPairedKeys()` since that indicates the Vector objects will be moved out of. The `platformCheckPairedKeys()` should just operate on normal const references.

javifernandez replied:
done

zdobersek commented on Source/WebCore/crypto/gcrypt/GCryptRFC7748.cpp:
```diff
@@ -0,0 +1,106 @@
+{
+
+    PAL::GCrypt::Handle<gcry_ctx_t> context;
+    if (auto error = gcry_mpi_ec_new(&context, nullptr, "Curve25519")) {
```
`gcry_mpi_ec_new` call should be using `ImplType::s_curveName`.

javifernandez replied:
done

zdobersek commented on Source/WebCore/crypto/gcrypt/GCryptRFC7748.cpp:
```diff
@@ -0,0 +1,106 @@
+template<typename ImplType>
+std::optional<Vector<uint8_t>> xImpl(const Vector<uint8_t>& kArg, const Vector<uint8_t>& uArg)
+
+{
```
Before the gcrypt context is created, the k and u Vectors should have their sizes validated. If they don't match the `argSize` for the given ImplType, this function should bail.

javifernandez replied:
done

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -397,6 +448,7 @@ Vector<uint8_t> CryptoKeyOKP::platformExportRaw() const
     return m_data;
 }
 
+
```
This newline isn't necessary.

javifernandez replied:
fixed

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoAlgorithmX25519GCrypt.cpp:
```diff
@@ -0,0 +1,49 @@
+
+namespace WebCore {
+
+static std::optional<Vector<uint8_t>> gcryptDerive(const Vector<uint8_t>& baseKey, const Vector<uint8_t>& publicKey, size_t)
```
This can be simplified to not receive an unused size parameter and also simply return the `X25519` call.

javifernandez replied:
I''ll do this refactoring in a follow up patch, thanks for the suggestion.

javifernandez replied:
Finally done in this PR

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -61,11 +67,11 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
 
     PAL::GCrypt::Handle<gcry_sexp_t> genkeySexp;
-    gcry_error_t error = gcry_sexp_build(&genkeySexp, nullptr, "(genkey (ecdsa (curve \"Ed25519\") (flags eddsa)))");
+    gcry_error_t error = gcry_sexp_build(&genkeySexp, nullptr, namedCurve == CryptoKeyOKP::NamedCurve::Ed25519 ? "(genkey (ecdsa (curve Ed25519) (flags eddsa)))" : "(genkey (ecc (curve Curve25519)))");
```
I think it would be cleaner, especially if Ed448 and X448 are supported at some point in the future, to disassociate the generation here into separate helper methods, already avoiding today's ternary operators.

For X25519, I would recommend employing the X25519 function on 32 bytes of random data, as per RFC7748, which would avoid these libgcrypt peculiarities regarding the leading byte.

All this can be discussed later and adjusted as necessary.

javifernandez replied:
I''ll do this refactoring in a follow up patch, thanks for the suggestion.

zdobersek reviewed (approved):
libgcrypt parts of the implementation look fine to me. @youennf please have a look at the common bits.

youennf commented on Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
```diff
@@ -216,6 +213,11 @@ std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmI
     return { };
 }
 
+bool CryptoKeyOKP::platformCheckPairedKeys(CryptoAlgorithmIdentifier, NamedCurve, Vector<uint8_t>&&, Vector<uint8_t>&&)
```
Should they be const& then?

javifernandez replied:
yeah, will submit a new patch fixing this. Thanks

javifernandez replied:
Done

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
```diff
@@ -0,0 +1,234 @@
+#include "CryptoKeyOKP.h"
+#include "ScriptExecutionContext.h"
+
+
```
Remove line?

javifernandez replied:
sure.

javifernandez replied:
done

youennf reviewed (approved):
I wonder whether we should have X25519 managed by its own runtime flag given safe curves flag is enabled by default.
If the GTK implementation is solid enough with this PR, that might be fine.

javifernandez committed:
[GTK] Implement the X25519 algorithm
https://bugs.webkit.org/show_bug.cgi?id=249017

Reviewed by Youenn Fablet and Žan Doberšek.

This change implements all the X25519's operations defined in the
WebCrypto Secure Curves specification. These operations include the
key generation, import/export in spki, pkcs8 and jwk formats and the
deriveKey and deriveBits methods that are the key operations of this
algorithm.

The main WebCrypto API implementation of the GTK+ port is based on
the Libgcrypt primitives. There is an ongoing effort to implement it
in OpenSSL (see bug 248980 for details) but it won't be an option for
the GTK+ port in the short term.

A new utility class has been added to the GCrypt based implementation
of the WebCrypto API, named as GCryptRFC7748. It implements the a function
to perform scalar multiplication on the Montgomery form of the Curve25519
and Curve488 curves. However, it only defines the X25519 function since
we don't support the X488 algorithm for now.

The feature is implemented behind a runtime flag, hence this PR adds a
new item in the UnifiedWebPreferences file as 'WebCryptoX25519Enabled'
with 'testable' status.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/cfrg_curves_bits.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/cfrg_curves_bits.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/cfrg_curves_keys.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/cfrg_curves_keys.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/failures_X25519.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/failures_X25519.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/successes_X25519.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/successes_X25519.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any.worker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_X25519.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_X25519.https.any.worker-expected.txt: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::write):
(WebCore::CloneDeserializer::read):
* Source/WebCore/crypto/CryptoAlgorithmIdentifier.h:
* Source/WebCore/crypto/CryptoAlgorithmParameters.h:
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::isX25519Enabled):
(WebCore::normalizeCryptoAlgorithmParameters):
(WebCore::isSupportedExportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp: Added.
(WebCore::CryptoAlgorithmX25519::create):
(WebCore::CryptoAlgorithmX25519::identifier const):
(WebCore::CryptoAlgorithmX25519::generateKey):
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):
(WebCore::CryptoAlgorithmX25519::deriveBits):
(WebCore::CryptoAlgorithmX25519::importKey):
(WebCore::CryptoAlgorithmX25519::exportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.h: Added.
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmRegistryGCrypt.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmX25519GCrypt.cpp: Added.
(WebCore::gcryptDerive):
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):
* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::CryptoKeyOKP::isPlatformSupportedCurve):
(WebCore::CryptoKeyOKPImpl::supportedAlgorithmIdentifier):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::exportSpki const):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::exportPkcs8 const):
(WebCore::CryptoKeyOKP::generateJwkX const):
* Source/WebCore/crypto/gcrypt/GCryptRFC7748.cpp: Added.
(WebCore::xImpl):
(WebCore::X25519):
* Source/WebCore/crypto/gcrypt/GCryptRFC7748.h: Added.
* Source/WebCore/crypto/gcrypt/GCryptUtilities.h:
* Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
(WebCore::CryptoKeyOKP::importRaw):
(WebCore::CryptoKeyOKP::importJwk):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::platformExportRaw const):
* Source/WebCore/crypto/keys/CryptoKeyOKP.h:
* Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
* Source/WebCore/crypto/parameters/CryptoAlgorithmX25519Params.h: Added.
* Source/WebCore/crypto/parameters/X25519Params.idl: Added.
* Source/WebCore/platform/SourcesGCrypt.txt:

Canonical link: https://commits.webkit.org/267602@main
